File: 0090-subsets-ii/0090-subsets-ii.cpp
class Solution {
private:
    void construct(vector<int> &nums, vector<vector<int>> &subsets, vector<int> &subset, int index) {

        subsets.push_back(subset);
        for (int i = index; i < nums.size(); i++) {
            if (i > index && nums[i] == nums[i - 1]) {
                continue;
            }

            subset.push_back(nums[i]);
            construct(nums, subsets, subset, i + 1);
            subset.pop_back();
        }
    }
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        vector<vector<int>> subsets;
        vector<int> subset;
        sort(nums.begin(), nums.end());

        construct(nums, subsets, subset, 0);
        
        return subsets;
    }
};

File: 0090-subsets-ii/0090-subsets-ii.java
class Solution {
    
    private void construct(int[] nums, List<List<Integer>> subsets, List<Integer> subset, int index) {
        subsets.add(List.copyOf(subset));

        for (int i = index; i < nums.length; i++) {
            if (i > index && nums[i] == nums[i - 1]) {
                continue;
            }

            subset.add(nums[i]);
            construct(nums, subsets, subset, i + 1); 
            subset.removeLast();
        }
    }
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        List<List<Integer>> subsets = new ArrayList<>();
        List<Integer> subset = new ArrayList<>();

        Arrays.sort(nums);
        this.construct(nums, subsets, subset, 0);

        return subsets;
    }
}